src/IB/IBFEMethod.cpp

```diff
@@ -693,50 +568,36 @@ IBFEMethod::preprocessIntegrateData(double current_time, double new_time, int nu
 void
 IBFEMethod::postprocessIntegrateData(double current_time, double new_time, int num_cycles)
 {
-    const std::string F_data_time_str = std::move(get_data_time_str(
-        d_ib_solver->getTimeSteppingType() == MIDPOINT_RULE ? d_half_time : d_new_time, d_current_time, d_new_time));
-    batch_vec_ghost_update(
-        { d_X_vecs->get("new"), d_U_vecs->get("new"), d_F_vecs->get(F_data_time_str), d_Q_half_vecs, d_Phi_half_vecs },
-        INSERT_VALUES,
-        SCATTER_FORWARD);
+    const std::string forcing_data_time_str = get_data_time_str(
+        d_ib_solver->getTimeSteppingType() == MIDPOINT_RULE ? d_half_time : d_new_time, d_current_time, d_new_time);
+    std::vector<std::vector<PetscVector<double>*> > vecs{ d_X_vecs->get("new"),
+                                                          d_U_vecs->get("new"),
+                                                          d_F_vecs->get(forcing_data_time_str) };
+    if (d_P_vecs) vecs.push_back(d_P_vecs->get(forcing_data_time_str));
+    if (d_Q_vecs) vecs.push_back(d_Q_vecs->get("half"));
+    batch_vec_ghost_update(vecs, INSERT_VALUES, SCATTER_FORWARD);
 
     d_X_vecs->copy("new", { "solution", "current" });
     d_U_vecs->copy("new", { "solution", "current" });
-    d_F_vecs->copy(F_data_time_str, { "solution", "current" });
+    d_F_vecs->copy(forcing_data_time_str, { "solution", "current" });
+    if (d_P_vecs) d_P_vecs->copy(forcing_data_time_str, { "solution", "current" });
+    if (d_Q_vecs) d_Q_vecs->copy("half", { "solution", "current" });
 
-    for (unsigned part = 0; part < d_num_parts; ++part)
+    // Update direct forcing data.
+    for (unsigned part = 0; part < d_meshes.size(); ++part)
     {
-        // Reset time-dependent Lagrangian data.
-        if (d_lag_body_source_part[part])
-        {
-            *d_Q_systems[part]->solution = *d_Q_half_vecs[part];
-            *d_Q_systems[part]->current_local_solution = *d_Q_half_vecs[part];
-        }
-
-        if (d_is_stress_normalization_part[part])
-        {
-            *d_Phi_systems[part]->solution = *d_Phi_half_vecs[part];
-            *d_Phi_systems[part]->current_local_solution = *d_Phi_half_vecs[part];
-        }
-
         if (d_direct_forcing_kinematics_data[part])
         {
             d_direct_forcing_kinematics_data[part]->postprocessIntegrateData(current_time, new_time, num_cycles);
         }
+    }
 
-        // Update the coordinate mapping dX = X - s.
+    // Update the coordinate mapping dX = X - s.
+    for (unsigned part = 0; part < d_meshes.size(); ++part)
+    {
```

Comment:
This doesn't seem right - we also call updateCoordinateMapping in `FEMechanicsBase::postprocessIntegrateData()`.

Reply:
In the initial implementation, `IBFEMethod::pre/postprocessIntegrateData()` did not call the base class implementation.

If you prefer to leverage the base class implementation, that's OK with me, but then I think the base class should not do any data allocation --- for instance, because subclasses could be using different time stepping schemes and need different system vector instances. (This already happens even within `IBFEMethod` depending on whether one uses midpoint or trapezoidal rule, for instance.)

Reply:
OK, so the base class now does essentially no data initialization or allocation. Once we have multiple implementations of the base class, perhaps we can think about whether it makes sense to factor some of that into the base class.